[FLINK-35022][Connector/DynamoDB] Add TypeInformed DDB Element Converter as default element converter

vahmed-hamdy opened the pull request:
<!--
*Thank you for contributing to Apache Flink AWS Connectors - we are happy that you want to help us improve our Flink connectors. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

- The name of the pull request should correspond to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
- Commits should be in the form of "[FLINK-XXXX][component] Title of the pull request", where [FLINK-XXXX] should be replaced by the actual issue number. 
    Generally, [component] should be the connector you are working on.
    For example: "[FLINK-XXXX][Connectors/Kinesis] XXXX" if you are working on the Kinesis connector or "[FLINK-XXXX][Connectors/AWS] XXXX" if you are working on components shared among all the connectors.
- Each pull request should only have one JIRA issue.
- Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

## Purpose of the change

Add `DynamoDbTypeInformedElementConverter` to convert Elements to dynamoDb Sink using its provided type Info.

## Verifying this change


This change added tests and can be verified as follows:

- Added unit tests

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [ ] New feature has been introduced
  - If yes, how is this documented? (not applicable / docs / JavaDocs / not documented)

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+    private <attrT> AttributeValue convertValue(
+            attrT attribute, TypeInformation<attrT> objectTypeInformation) {
+        if (attribute == null) {
+            return AttributeValue.builder().nul(true).build();
```
Not sure this is the correct thing to do here? I believe a common practise is to omit null fields rather than setting to null. Maybe this could be a configuration on the element converter?

vahmed-hamdy replied:
@dannycranmer , yeah just checked `testConvertOrderToDynamoDbWriteRequestWithIgnoresNull`.

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+        for (String fieldKey : typeInfo.getFieldNames()) {
+            TypeInformation<attT> fieldType = typeInfo.getTypeAt(fieldKey);
+            try {
+                Field field = t.getClass().getDeclaredField(fieldKey);
```
As per the [coding guidelines](https://flink.apache.org/how-to-contribute/code-style-and-quality-java/#java-reflection) we should not use reflection. Can we instead use the `TypeExtractor` class as suggested in the docs? 

Another note here is that loading the class information for each record is expensive. Suggest that we lazily cache the type info based on the class.

dannycranmer commented:
Please change commit message to include the component: `[FLINK-35022][Connectors/DynamoDB]`

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+        } else if (componentTypeInfo instanceof NumericTypeInfo) {
+            if (attribute instanceof int[]) {
+                return AttributeValue.fromNs(
+                        (Arrays.stream((int[]) attribute)
```
As a big fan of the stream API, it pains me to say that we [should not be using it for performance critical code](https://flink.apache.org/how-to-contribute/code-style-and-quality-java/#java-streams). Since this would be called for each record I would consider this performance critical/data-intensive. Please convert to standard loops, can extract to new methods for readability.

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverterTest.java:
```diff
@@ -0,0 +1,161 @@
+public class DynamoDbTypeInformedElementConverterTest {
+
+    @Test
+    public void testSimpleTypeConversion() {
```
nit: JUnit 5 coding standards are slightly different from lower versions. In 5 we:
  1/ Use package protected class/methods (drop `public`)
  2/ Remove the `test` prefix on the methods (`testSimpleTypeConversion()` -> `simpleTypeConversion()`)

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/test/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverterTest.java:
```diff
@@ -0,0 +1,161 @@
+
+        Date dateVal = Date.from(Instant.now());
+
+        Throwable throwable =
+                Assertions.assertThrows(
+                        FlinkRuntimeException.class,
+                        () -> {
+                            elementConverter.apply(Tuple2.of(dateVal, "stringVal"), null);
+                        });
+
+        assertThat(throwable).hasCauseInstanceOf(IllegalArgumentException.class);
+        assertThat(throwable.getCause())
+                .hasMessageContaining("Unsupported BasicTypeInfo " + BasicTypeInfo.DATE_TYPE_INFO);
```
Can you use the AssertJ lib here instead?

```
Assertions.assertThatExceptionOfType(FlinkRuntimeException.class)
                .isThrownBy(() -> elementConverter.apply(Tuple1.of("nan"), null))
                .withMessageContaining("Incompatible type when converting nan, using Integer type information.")
                .withRootCauseInstanceOf(IllegalArgumentException.class);
```

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+                                .boxed()
+                                .map(Object::toString)
+                                .collect(Collectors.toList())));
+            } else if (attribute instanceof double[]) {
+                return AttributeValue.fromNs(
+                        (Arrays.stream((double[]) attribute)
+                                .boxed()
+                                .map(Object::toString)
+                                .collect(Collectors.toList())));
+            } else if (attribute instanceof long[]) {
+                return AttributeValue.fromNs(
+                        (Arrays.stream((long[]) attribute)
+                                .boxed()
+                                .map(Object::toString)
+                                .collect(Collectors.toList())));
+            } else {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Unsupported primitive numeric typeInfo %s",
+                                componentTypeInfo.getTypeClass()));
+            }
+        } else if (!isPrimitive) {
+            List<AttributeValue> attributeValueList = new ArrayList<>();
+            for (Object attributeItem : (Object[]) attribute) {
+                checkTypeCompatibility(attributeItem, componentTypeInfo);
+                attributeValueList.add(convertValue(((attrT) attributeItem), componentTypeInfo));
+            }
+            return AttributeValue.fromL(attributeValueList);
+        } else {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Unsupported primitive typeInfo %s", componentTypeInfo.getTypeClass()));
+        }
```
This block is untested, please add tests

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+                    componentTypeInfo,
+                    objectTypeInformation instanceof PrimitiveArrayTypeInfo);
+
+        } else if (objectTypeInformation instanceof TupleTypeInfo) {
+            TupleTypeInfo<?> tupleTypeInfo = (TupleTypeInfo<?>) objectTypeInformation;
+            Tuple t = (Tuple) attribute;
+            List<AttributeValue> attributeValues = new ArrayList<>();
+            for (int i = 0; i < tupleTypeInfo.getTotalFields(); i++) {
+                attributeValues.add(convertValue(t.getField(i), tupleTypeInfo.getTypeAt(i)));
+            }
+            return AttributeValue.fromL(attributeValues);
```
This block is untested, please add tests

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,245 @@
+            // special handling for byte array as it has special support in DynamoDB
+            if (attribute instanceof byte[]) {
+                return AttributeValue.fromB(SdkBytes.fromByteArray((byte[]) attribute));
+            } else if (attribute instanceof SdkBytes) {
+                return AttributeValue.fromB((SdkBytes) attribute);
+            } else {
+                throw new IllegalArgumentException(
+                        String.format("Failed to extract byte array type from %s", attribute));
+            }
```
This block is untested, please add tests

dannycranmer commented:
I have left a comment regarding the need for this on the Jira: https://issues.apache.org/jira/browse/FLINK-35022

vahmed-hamdy commented:
@dannycranmer Thanks for the feedback, I have addressed your comments. Alternatively I decided to follow an approach similar to `RowDataToAttributeValueConverter` to avoid using reflection. PTAL

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,380 @@
+
+    @Override
+    public void open(Sink.InitContext context) {
+        try {
+            tableSchema = createTableSchema(typeInfo);
+        } catch (IntrospectionException | IllegalStateException | IllegalArgumentException e) {
+            throw new FlinkRuntimeException("Failed to extract DynamoDb table schema", e);
+        }
```
We should do this in the constructor as a validation and fail fast. Do not need to store, but nice to fail on client and not when running job

dannycranmer commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,380 @@
+
+    private <attrT> TableSchema<attrT> createTableSchemaFromPojo(PojoTypeInfo<attrT> typeInfo)
+            throws IntrospectionException {
+        StaticTableSchema.Builder<attrT> tableSchemaBuilder =
+                StaticTableSchema.builder(typeInfo.getTypeClass());
+        BeanInfo beanInfo = Introspector.getBeanInfo(typeInfo.getTypeClass());
+        PropertyDescriptor[] propertyDescriptors = beanInfo.getPropertyDescriptors();
```
This looks like it might be expensive, can we cache the descriptors per type?

vahmed-hamdy replied:
I am not sure I get what we need to cache, these descriptors are only created once when creating the schema, during element conversion only the Enhanced getters are used.
Btw, this is also what the BeanTableSchema does :D

dannycranmer reviewed (changes requested):
Thanks for the update @vahmed-hamdy. Left a few comments

dannycranmer commented:
@vahmed-hamdy can you also update the docs please?

vahmed-hamdy commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,380 @@
+ * TypeInformation provided.
+ */
+@PublicEvolving
+public class DynamoDbTypeInformedElementConverter<inputT>
```
I agree, I would suggest `DefaultDynamoDbElementConverter` since we are following up with making it default and changing it to optional? wdyt?

dannycranmer reviewed (approved):
LGTM, thanks @vahmed-hamdy

dannycranmer commented:
@hlteoh37 / @z3d1k can you please take a look at this one too?

hlteoh37 commented on flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java:
```diff
@@ -0,0 +1,380 @@
+                    .converterFor(EnhancedType.of(typeInfo.getTypeClass()));
+        } else if (typeInfo.equals(PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)) {
+            return getAttributeConverter(
+                    AttributeValueType.B,
+                    bytes ->
+                            bytes instanceof SdkBytes
+                                    ? AttributeValue.fromB((SdkBytes) bytes)
+                                    : AttributeValue.fromB(SdkBytes.fromByteArray((byte[]) bytes)));
+        } else if (typeInfo instanceof BasicArrayTypeInfo) {
+            BasicArrayTypeInfo<AttributeT, ?> basicArrayTypeInfo =
+                    (BasicArrayTypeInfo<AttributeT, ?>) typeInfo;
+            if (basicArrayTypeInfo.getComponentInfo().equals(BasicTypeInfo.STRING_TYPE_INFO)) {
+                return getAttributeConverter(
+                        AttributeValueType.SS,
+                        array -> AttributeValue.fromSs(Arrays.asList((String[]) array)));
+            } else if (basicArrayTypeInfo.getComponentInfo() instanceof NumericTypeInfo) {
+                return getAttributeConverter(
+                        AttributeValueType.NS,
+                        array ->
+                                AttributeValue.fromNs(
+                                        convertObjectArrayToStringList((Object[]) array)));
+            }
+
+            return new ArrayAttributeConverterProvider()
+                    .converterFor(EnhancedType.of(typeInfo.getTypeClass()));
+        } else if (typeInfo instanceof ObjectArrayTypeInfo) {
+            return getObjectArrayTypeConverter((ObjectArrayTypeInfo<AttributeT, ?>) typeInfo);
+        } else if (typeInfo instanceof PrimitiveArrayTypeInfo) {
+            PrimitiveArrayTypeInfo<AttributeT> primitiveArrayTypeInfo =
+                    (PrimitiveArrayTypeInfo<AttributeT>) typeInfo;
+            if (primitiveArrayTypeInfo.getComponentType() instanceof NumericTypeInfo) {
+                return getAttributeConverter(
+                        AttributeValueType.NS,
+                        array -> AttributeValue.fromNs(convertPrimitiveArrayToStringList(array)));
+            } else {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Unsupported primitive array typeInfo %s",
+                                primitiveArrayTypeInfo.getComponentType()));
+            }
```
Have we considered using the `.attributeConverterProviders()` method in the table schema builder and reusing the constructs we use to DDB Table API?

```
TableSchema.builder(...)
                      .attributeConverterProviders(new ArrayAttributeConverterProvider(), ... (add any more here) ..., AttributeConverterProvider.defaultProvider());
```

We have an example here:
https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java#L56-L72

vahmed-hamdy replied:
@hlteoh37 yes, the whole approach actually is very similar, however The way `AttributeConverterProvider` works is by mapping `EnhancedType` provided to an `AttributeConverter`, We here rely on Flink's `TypeInfo` rather than the `EnhancedType` that's why we created an alternative structure. Wdyt?

hlteoh37 replied:
Hm, ultimately `EnhancedType` is just a wrapper for the `AttributeConverterProvider` to match the right `AttributeConverter`. We cannot run away from this - in fact you have created a custom implementation of this logic in `getAttributeConverter`, which takes type information and constructs a custom attribute converter. 

Wouldn't using the `AttributeConverterProvider` more directly simplify the code?

vahmed-hamdy replied:
> We cannot run away from this 

This is not true, we are "running away from this" or rather replacing the mapping dependency on `EnhancedType` by directly attaching each resolved `AttribureConverter` to the attribute in
```
  typeInfo.getTypeClass(),
                a ->
                        a.name(fieldName)
                                .getter(getter)
                                .setter((e, o) -> {})
                                .attributeConverter(getAttributeConverter(typeInfo)))
```

> in fact you have created a custom implementation of this logic in getAttributeConverter, which takes type information and constructs a custom attribute converter.

Yes, I am aware of that and that exactly why using an attribute converter is against what we are trying to achieve here, Are you suggesting we don't depend on Flink's TypeInformation and resolve back to Java Classes?

hlteoh37 reviewed (changes requested):
Thanks for the PR @vahmed-hamdy ! Added a comment

vahmed-hamdy commented:
@hlteoh37 Thanks, for the feedback, the reason we took this approach is that We are trying to couple it as much as possible with Flink's TypeInfo Class, Using `AttributeConverterProvider` is closer to DDB's `EnhancedType` rather than Flink's

vahmed-hamdy commented:
@hlteoh37 Do you have other comments or could we merge this PR?

hlteoh37 commented:
Hi @vahmed-hamdy, my preference would be for us to refactor the code in `DynamoDbTypeInformedElementConverter` to use a cleaner structure rather than a long list of `if-else`. Is this something we need to close urgently?

hlteoh37 reviewed (approved):
Thanks @vahmed-hamdy
